packages/graphql/src/services/notifications.ts

```diff
@@ -49,3 +51,130 @@ export async function subscribeToPush(
     },
   })
 }
+
+interface UnsubscribeToPushArgs {
+  courseId: string
+  endpoint: string
+}
+
+export async function unsubscribeFromPush( 
+  { courseId, endpoint }: UnsubscribeToPushArgs,
+  ctx: ContextWithUser
+) {
+  try {
+    await ctx.prisma.pushSubscription.delete({
+      where: {
+        participantId_courseId_endpoint: {
+          participantId: ctx.user.sub,
+          courseId,
+          endpoint
+        }
+      }
+    })
+    return true
+  } catch(error) {
+    console.error("An error occured while trying to unsubscribe from push notifications: ", error)
+    return false
+  }
+}
+
+export async function sendPushNotifications( ctx: Context) {
+  webpush.setVapidDetails(
+    'mailto:klicker.support@uzh.ch',
+    process.env.VAPID_PUBLIC_KEY as string,
+    process.env.VAPID_PRIVATE_KEY as string
+  )
+
+  const microSessions = await ctx.prisma.microSession.findMany({
+    where: {
+      status: MicroSessionStatus.PUBLISHED,
+      scheduledStartAt: {
+        lte: new Date(),
+      },
+      scheduledEndAt: {
+        gt: new Date(),
+      },
+      arePushNotificationsSent: false,
+    },
+    include: {
+      course: {
+        include: {
+          subscriptions: true,
+        },
+      },
+      instances: false,
+    },
+  })
+
+
+  await Promise.all(
+    microSessions.map(async (microSession: MicroSession) => {
+      try {
+        await sendPushNotificationsToSubscribers(microSession.course.subscriptions, ctx);
+        
+        //update microSession to prevent sending push notifications multiple times
+        const updatedMicroSession = await ctx.prisma.microSession.update({
+          where: {
+            id: microSession.id,
+          },
+          data: {
+            arePushNotificationsSent: true,
+          },
+        })
+      } catch (error) {
+        console.error("An error occured while trying to send the push notifications: ", error)
+      }
+    })
+ )
+
+  return true
+}
+
+function sleep(ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms))
+}
+
+async function sendPushNotificationsToSubscribers(subscriptions: PushSubscription[], ctx: Context) {
+  for (let sub of subscriptions) {
+    await sleep(500)
+    const subscription = {
+      endpoint: sub.endpoint,
+      keys: {
+        auth: sub.auth,
+        p256dh: sub.p256dh,
+      },
+    }
+
+    try {
+      await webpush.sendNotification(
+        {
+          endpoint: sub.endpoint,
+          keys: {
+            auth: sub.auth,
+            p256dh: sub.p256dh,
+          },
+        },
+        JSON.stringify({
+          message:
+            'Das Microlearning für BFI Woche 5 ist bis morgen um 09:00 verfügbar.',
```

Comment:
the message needs to be parametrized to reflect the name of the course and the name of the microlearning, as well as the timing. anything else that the lecturer should be able to customize?